### PR TITLE
GEODE-8656: Fix ping only sent to one gateway receiver when several r…

### PR DESCRIPTION
**Pull request by albertogpz**
…eceivers share the same hostname-for-senders

When serveral receivers have the same hostname-for-senders only one of them received
pings.
The reason was that the structure holding the endpoints was a Map whose key was
the server location. As all the endpoints for the remote gateway senders shared
the same server location, only one received the ping.

The structure holding endpoints in the endpoint manager has been updated to use
as key a ServerLocationAndMemberId class instead of ServerLocation.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Review by kohlmu-pivotal — CHANGES_REQUESTED** [review]
Hi there @albertogpz.
Thank you for this contribution, but would it be possible to create a DUnit test to test out the functionality that you are providing.

DUnit tests, is our current way to spin up multiple Servers in different VM's in order to test out multi-server functionality.

**Comment by albertogpz** [comment]
> Hi there @albertogpz.
> Thank you for this contribution, but would it be possible to create a DUnit test to test out the functionality that you are providing.
> 
> DUnit tests, is our current way to spin up multiple Servers in different VM's in order to test out multi-server functionality.

Hi @kohlmu-pivotal .
I thought about creating a DUnit test for this change but the problem I ran into is that it would require to have a load balancer that balances connections between the servers that share the same hostname-for-senders value.
Any idea how to implement this with the current framework?

**Comment by kohlmu-pivotal** [comment]
@albertogpz, hostname-for-senders is a string value which represents the hostname that is given to the senders. This means, that if you start 2 (or more servers) on the same machine they can have the same value. Could you create two servers with a gateway receiver each, with the same hostname-for-server value and see if this solves the problem.

If not, we have to seriously put our heads together and try and come up with a solution to the problem. As I'd prefer a test to be in place that reproduce the issue and then of course fix it. Might even mean that we might have to create a whole new framework to test these kind of issues.

**Comment by albertogpz** [comment]
> @albertogpz, hostname-for-senders is a string value which represents the hostname that is given to the senders. This means, that if you start 2 (or more servers) on the same machine they can have the same value. Could you create two servers with a gateway receiver each, with the same hostname-for-server value and see if this solves the problem.
> 
> If not, we have to seriously put our heads together and try and come up with a solution to the problem. As I'd prefer a test to be in place that reproduce the issue and then of course fix it. Might even mean that we might have to create a whole new framework to test these kind of issues.

@kohlmu-pivotal, the receivers part is easy, we can create two senders with the same hostname-for-senders value as you suggest.
The difficult part is the sending side. The gateway senders would use the value of hostname-for-senders to connect to the receivers but we should have "something" (the load balancer in the real world) that should intercept the connect call to the hostname-for-senders address so that the connection is finally made against the ip addresses to which the receivers binded.
I do not know if we can tweak in the test case the part of the creation of the connection in the gateway sender so that when a connection to an endpoint is created, the IP address to be used is changed to one of the IP addresses on which the receivers are listening.
I will try to check if this is possible.
Any ideas are welcome.
Thanks!

**Comment by kohlmu-pivotal** [comment]
@albertogpz some engineers pointed me at `DualServerSNIAcceptanceTest` to possibly be a template to help with testing this.
I will admit I'm not across anything done in this test, maybe @Bill could provide some guidance if you have more questions

**Comment by albertogpz** [comment]
> @albertogpz some engineers pointed me at `DualServerSNIAcceptanceTest` to possibly be a template to help with testing this.
> I will admit I'm not across anything done in this test, maybe @Bill could provide some guidance if you have more questions

@kohlmu-pivotal Great, thanks! I will take a look at that test and if @Bill wants to comment, he is more than welcome.

**Comment by albertogpz** [comment]
> > @albertogpz some engineers pointed me at `DualServerSNIAcceptanceTest` to possibly be a template to help with testing this.
> > I will admit I'm not across anything done in this test, maybe @Bill could provide some guidance if you have more questions
> 
> @kohlmu-pivotal Great, thanks! I will take a look at that test and if @Bill wants to comment, he is more than welcome.

I created a new distributed test by looking at what was done in the test you pointed me to but I have just realized that docker-compose is not available in the machines where the distributed tests are running.
@Bill, @kohlmu-pivotal , how can I make docker-compose be available for distributed tests? Or should I move the test to the acceptance tests ones?

**Comment by Bill** [comment]
@albertogpz if you want to use Docker in your tests, yeah, the easiest way to do that right now is to make your test an "acceptanceTest" which means putting it in an `/geode-assembly/src/acceptanceTest` hierarchy.

**Comment by albertogpz** [comment]
> @albertogpz if you want to use Docker in your tests, yeah, the easiest way to do that right now is to make your test an "acceptanceTest" which means putting it in an `/geode-assembly/src/acceptanceTest` hierarchy.

Thanks, @Bill . How hard would it be to use docker-compose in a distributedTest? I am saying it because I used in my test the org.apache.geode.test.dunit.VM class. If I move it to be an acceptanceTest I would have to implement it in a different way. It looks to me that in acceptanceTest, Geode processes are started with gfsh. Any suggestion?

**Comment by albertogpz** [comment]
> > @albertogpz if you want to use Docker in your tests, yeah, the easiest way to do that right now is to make your test an "acceptanceTest" which means putting it in an `/geode-assembly/src/acceptanceTest` hierarchy.
> 
> Thanks, @Bill . How hard would it be to use docker-compose in a distributedTest? I am saying it because I used in my test the org.apache.geode.test.dunit.VM class. If I move it to be an acceptanceTest I would have to implement it in a different way. It looks to me that in acceptanceTest, Geode processes are started with gfsh. Any suggestion?

I finally created the test as acceptanceTest.

**Comment by kohlmu-pivotal** [comment]
@albertogpz, I'm truly not the best SME to evaluate this PR. I have requested (and added) someone onto this PR, who will hopefully review this PR for correctness

**Comment by albertogpz** [comment]
> @albertogpz, I'm truly not the best SME to evaluate this PR. I have requested (and added) someone onto this PR, who will hopefully review this PR for correctness

Thanks @kohlmu-pivotal . If you could add @bschuchardt it would probably be best as he reviewed the previous tickets related to this issue.

**Review by bschuchardt — CHANGES_REQUESTED** [review]
Nice work Alberto.  I appreciate that you persevered and added a Dockerized test that exhibits your achievements.

I wonder if you'd be willing to write something up to contribute to Geode documentation about this use case.  Other people could benefit from knowing about this option.

I do have a couple of minor changes that I'd like to see in the new acceptance test.

File: geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/wan/SeveralGatewayReceiversWithSamePortAndHostnameForSenders.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    createPartitionedRegion(vm1, regionName, senderId, 0, 10);
+
+    int NUM_PUTS = 10;
+
+    putKeyValues(vm1, NUM_PUTS, regionName);
+
+    // Wait longer than the value set in the receivers for
+    // maximum-time-between-pings (60000) to verify that connections are not closed
+    // by the receivers because each has received the pings timely.
+    int maxTimeBetweenPingsInReceiver = 65000;
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/wan/SeveralGatewayReceiversWithSamePortAndHostnameForSenders.java]: It was not possible before the fix for #5701 but as that is now in place, I have been able to reduce the max to 15 secs.

**Comment by albertogpz** [comment]
> Nice work Alberto. I appreciate that you persevered and added a Dockerized test that exhibits your achievements.
> 
> I wonder if you'd be willing to write something up to contribute to Geode documentation about this use case. Other people could benefit from knowing about this option.
> 
> I do have a couple of minor changes that I'd like to see in the new acceptance test.

Thanks @bschuchardt . I agree it would be good to write something in the documentation about this use case. I assume you are talking about WAN deployments in which the topology of the Geode cluster (the servers) is hidden behind a single IP address. Am I correct?

If so, I will look for a place in the Geode documentation to add to it. I think it will be best to do it in a new PR.

**Comment by albertogpz** [comment]
> > Nice work Alberto. I appreciate that you persevered and added a Dockerized test that exhibits your achievements.
> > I wonder if you'd be willing to write something up to contribute to Geode documentation about this use case. Other people could benefit from knowing about this option.
> > I do have a couple of minor changes that I'd like to see in the new acceptance test.
> 
> Thanks @bschuchardt . I agree it would be good to write something in the documentation about this use case. I assume you are talking about WAN deployments in which the topology of the Geode cluster (the servers) is hidden behind a single IP address. Am I correct?
> 
> If so, I will look for a place in the Geode documentation to add to it. I think it will be best to do it in a new PR.

@bschuchardt I have created a new ticket GEODE-8738 and a PR #5766 with the information you suggested.

**Review by bschuchardt — APPROVED** [review]

**Review by kohlmu-pivotal — APPROVED** [review]

**Pull request merged**